File: packages/fluxguard-node/src/FluxGuard.ts
import {
  checkFixedWindowLocal,
  checkSlidingWindowCounterLocal,
  checkSlidingWindowLogLocal,
  checkTokenBucketLocal,
} from "./engine/localChecks.js";
import { runRedisCheck } from "./engine/redisChecks.js";
import { LocalStore } from "./store/LocalStore.js";
import { RedisStore } from "./store/RedisStore.js";
import type { Store } from "./store/Store.js";
import {
  Algorithm,
  type FluxGuardConfig,
  type FluxGuardMetricsSnapshot,
  type RateLimitResult,
} from "./types.js";

export class FluxGuard {
  private readonly store: Store;
  private readonly prefix: string;
  private readonly nowFn: () => number;
  private readonly config: FluxGuardConfig;
  private metrics: FluxGuardMetricsSnapshot = {
    totalChecks: 0,
    allowed: 0,
    throttled: 0,
    redisErrors: 0,
  };

  constructor(private readonly cfg: FluxGuardConfig) {
    this.config = cfg;
    this.prefix = cfg.keyPrefix ?? "fluxguard:";
    this.nowFn = cfg.nowFn ?? (() => Date.now());
    if (cfg.redis) {
      const r = cfg.redis as object;
      const isClient = "connect" in r && typeof (r as { connect?: unknown }).connect === "function";
      this.store = new RedisStore(cfg.redis, !isClient);
    } else {
      this.store = new LocalStore(this.nowFn);
    }
  }

  getMetrics(): FluxGuardMetricsSnapshot {
    return { ...this.metrics };
  }

  /** Merge event recording for external metrics (e.g. Prometheus). */
  onMetricsRecord(fn: (snapshot: FluxGuardMetricsSnapshot) => void): void {
    const prev = this.config.metrics?.record;
    this.config.metrics = {
      ...this.config.metrics,
      record: (s) => {
        prev?.(s);
        fn(s);
      },
    };
  }

  async check(key: string): Promise<RateLimitResult> {
    this.metrics.totalChecks += 1;
    try {
      let result: RateLimitResult;
      if (this.store instanceof RedisStore) {
        result = await runRedisCheck(
          this.store,
          this.prefix,
          key,
          this.cfg.algorithm,
          this.cfg.limit,
          this.cfg.windowMs,
          this.nowFn(),
        );
      } else {
        result = await this.checkLocal(key);
      }
      if (result.allowed) {
        this.metrics.allowed += 1;
        this.config.events?.onAllowed?.(key, result);
      } else {
        this.metrics.throttled += 1;
        this.config.events?.onThrottled?.(key, result);
      }
      this.config.metrics?.record?.(this.getMetrics());
      return result;
    } catch (err) {
      if (this.store instanceof RedisStore) {
        this.metrics.redisErrors += 1;
        this.config.events?.onRedisError?.(err);
        if (this.cfg.failOpen !== false) {
          const now = this.nowFn();
          const allowed: RateLimitResult = {
            allowed: true,
            limit: this.cfg.limit,
            remaining: this.cfg.limit,
            resetMs: now + this.cfg.windowMs,
            algorithm: this.cfg.algorithm,
          };
          this.metrics.allowed += 1;
          this.config.metrics?.record?.(this.getMetrics());
          return allowed;
        }
      }
      throw err;
    }
  }

  private async checkLocal(key: string): Promise<RateLimitResult> {
    const now = this.nowFn();
    const { algorithm, limit, windowMs } = this.cfg;
    switch (algorithm) {
      case Algorithm.FIXED_WINDOW:
        return checkFixedWindowLocal(this.store, this.prefix, key, limit, windowMs, now);
      case Algorithm.TOKEN_BUCKET:
        return checkTokenBucketLocal(this.store, this.prefix, key, limit, windowMs, now);
      case Algorithm.SLIDING_WINDOW_LOG:
        return checkSlidingWindowLogLocal(this.store, this.prefix, key, limit, windowMs, now);
      case Algorithm.SLIDING_WINDOW_COUNTER:
        return checkSlidingWindowCounterLocal(this.store, this.prefix, key, limit, windowMs, now);
      default:
        throw new Error(`Unsupported algorithm: ${String(algorithm)}`);
    }
  }
}

File: packages/fluxguard-java/src/main/java/com/fluxguard/core/FluxGuard.java
package com.fluxguard.core;

import com.fluxguard.engine.LocalAlgorithms;
import com.fluxguard.engine.RedisAlgorithms;
import com.fluxguard.store.LocalStore;
import com.fluxguard.store.RedisStore;
import com.fluxguard.store.Store;
import io.lettuce.core.RedisClient;
import io.lettuce.core.api.StatefulRedisConnection;

public final class FluxGuard implements AutoCloseable {
  private final FluxGuardConfig config;
  private final Store store;
  private final RedisStore redisStore;
  private final boolean useRedis;
  private final java.util.function.LongSupplier nowFn;
  private final String prefix;
  private final StatefulRedisConnection<String, String> redisConnection;
  private final RedisClient redisClient;

  private long totalChecks;
  private long allowed;
  private long throttled;
  private long redisErrors;

  public FluxGuard(FluxGuardConfig config) {
    this.config = config;
    this.prefix = config.keyPrefix();
    this.nowFn = config.nowFn();
    if (config.redisClient() != null) {
      this.redisClient = config.redisClient();
      this.redisConnection = redisClient.connect();
      this.redisStore = new RedisStore(redisConnection.sync());
      this.store = redisStore;
      this.useRedis = true;
    } else {
      this.redisClient = null;
      this.redisConnection = null;
      this.redisStore = null;
      this.store = new LocalStore(config.nowFn());
      this.useRedis = false;
    }
  }

  @Override
  public void close() {
    if (redisConnection != null) {
      redisConnection.close();
    }
    if (redisClient != null) {
      redisClient.shutdown();
    }
  }

  public RateLimitResult check(String key) {
    totalChecks++;
    long now = nowFn.getAsLong();
    try {
      RateLimitResult result;
      if (useRedis) {
        result =
            RedisAlgorithms.run(
                redisStore,
                prefix,
                key,
                config.algorithm(),
                config.limit(),
                config.windowMs(),
                now);
      } else {
        result = checkLocal(key, now);
      }
      if (result.allowed()) {
        allowed++;
      } else {
        throttled++;
      }
      return result;
    } catch (Exception e) {
      if (useRedis) {
        redisErrors++;
        if (config.failOpen()) {
          allowed++;
          return new RateLimitResult(
              true,
              config.limit(),
              config.limit(),
              now + config.windowMs(),
              java.util.OptionalLong.empty(),
              config.algorithm());
        }
      }
      throw new RuntimeException(e);
    }
  }

  private RateLimitResult checkLocal(String key, long now) {
    int limit = config.limit();
    long windowMs = config.windowMs();
    Algorithm a = config.algorithm();
    if (a == Algorithm.FIXED_WINDOW) {
      return LocalAlgorithms.fixedWindow(store, prefix, key, limit, windowMs, now);
    }
    if (a == Algorithm.TOKEN_BUCKET) {
      return LocalAlgorithms.tokenBucket(store, prefix, key, limit, windowMs, now);
    }
    if (a == Algorithm.SLIDING_WINDOW_LOG) {
      return LocalAlgorithms.slidingWindowLog(store, prefix, key, limit, windowMs, now);
    }
    if (a == Algorithm.SLIDING_WINDOW_COUNTER) {
      return LocalAlgorithms.slidingWindowCounter(store, prefix, key, limit, windowMs, now);
    }
    throw new IllegalStateException("Unhandled algorithm: " + a);
  }

  public long getTotalChecks() {
    return totalChecks;
  }

  public long getAllowed() {
    return allowed;
  }

  public long getThrottled() {
    return throttled;
  }

  public long getRedisErrors() {
    return redisErrors;
  }
}
